compiler/rustc_target/src/spec/mod.rs

```diff
@@ -2849,6 +2850,303 @@ impl Target {
+            );
+        }
+
+        if self.link_self_contained.is_disabled() {
+            check!(
+                self.pre_link_objects_self_contained.is_empty()
+                    && self.post_link_objects_self_contained.is_empty(),
+                "if `link_self_contained` is disabled, then `pre_link_objects_self_contained` and `post_link_objects_self_contained` must be empty",
+            );
+        }
+
+        // If your target really needs to deviate from the rules below,
+        // except it and document the reasons.
+        // Keep the default "unknown" vendor instead.
+        check_ne!(self.vendor, "", "`vendor` cannot be empty");
+        check_ne!(self.os, "", "`os` cannot be empty");
+        if !self.can_use_os_unknown() {
+            // Keep the default "none" for bare metal targets instead.
+            check_ne!(
+                self.os,
+                "unknown",
+                "`unknown` os can only be used on particular targets; use `none` for bare-metal targets"
+            );
+        }
+
+        // Check dynamic linking stuff
+        // BPF: when targeting user space vms (like rbpf), those can load dynamic libraries.
+        // hexagon: when targeting QuRT, that OS can load dynamic libraries.
+        // wasm{32,64}: dynamic linking is inherent in the definition of the VM.
+        if self.os == "none"
+            && (self.arch != "bpf"
+                && self.arch != "hexagon"
+                && self.arch != "wasm32"
+                && self.arch != "wasm64")
+        {
+            check!(
+                !self.dynamic_linking,
+                "dynamic linking is not supported on this OS/architecture"
+            );
+        }
+        if self.only_cdylib
+            || self.crt_static_allows_dylibs
+            || !self.late_link_args_dynamic.is_empty()
+        {
+            check!(
+                self.dynamic_linking,
+                "dynamic linking must be allowed when `only_cdylib` or `crt_static_allows_dylibs` or `late_link_args_dynamic` are set"
+            );
+        }
+        // Apparently PIC was slow on wasm at some point, see comments in wasm_base.rs
+        if self.dynamic_linking && !(self.is_like_wasm && self.os != "emscripten") {
+            check_eq!(
+                self.relocation_model,
+                RelocModel::Pic,
+                "targets that support dynamic linking must use the `pic` relocation model"
+            );
+        }
+        if self.position_independent_executables {
+            check_eq!(
+                self.relocation_model,
+                RelocModel::Pic,
+                "targets that support position-independent executables must use the `pic` relocation model"
+            );
+        }
+        // The UEFI targets do not support dynamic linking but still require PIC (#101377).
+        // We also allow this for JSON targets since otherwise, our default values would fail this test.
```

Comment:
this feels questionable :sweat_smile: 

maybe change this to a FIXME :thinking: it feels suboptimal/dangerous for our default target values to be incorrect

Reply:
Added a FIXME an filed an issue: https://github.com/rust-lang/rust/issues/133459